lib/benchee/project/benchmark.ex

```diff
@@ -0,0 +1,175 @@
+      end
+    end
+  end
+
+  @doc """
+  Module-scope teardown. Optional
+
+  Can return anything.
+  """
+  defmacro teardown(_opts \\ [], do: doblock) do
+    quote do
+      def teardown(state) do
+        unquote(doblock)
+      end
+    end
+  end
+
+  # credo:disable-for-next-line Credo.Check.Refactor.CyclomaticComplexity
+  defmacro benchmark(name, opts \\ [], do: {:__block__, [], exprs}) do
+    mod = __CALLER__.module
+    state_var = atom_to_var(:state)
+
+    inputs =
+      Keyword.get(opts, :inputs, []) ++
+        for {:input, _, [name, expr]} <- exprs,
+            do:
+              {name,
+               quote do
+                 unquote(block_to_fun(expr)).()
+               end}
+
+    case_args =
+      case length(inputs) do
+        0 -> []
+        _ -> [:input]
+      end
+
+    scenarios =
+      for {:scenario, _, args} <- exprs do
+        case args do
+          [name, block] ->
+            {name, block_to_fun(block, case_args)}
+
+          [name, opts, block] ->
+            {name, {block_to_fun(block, case_args), opts}}
+        end
+      end
+
+    [setup_doblock | rest] =
+      (exprs
+       |> Enum.filter(fn t -> elem(t, 0) == :setup end)
+       |> Enum.map(fn {:setup, _, [[do: doblock]]} -> doblock end)) ++
+        [
+          quote generated: true do
+            {:ok, unquote(state_var)}
+          end
+        ]
+
+    local_setup = block_to_fun(setup_doblock, [:state])
+
+    if length(rest) > 1 do
+      :ok = Logger.warn("#{inspect(mod)}: Only the first setup is used")
+    end
+
+    [teardown_doblock | rest] =
+      (exprs
+       |> Enum.filter(fn t -> elem(t, 0) == :teardown end)
+       |> Enum.map(fn {:teardown, _, [[do: doblock]]} -> doblock end)) ++
+        [
+          quote do
+            {:ok, unquote(state_var)}
+          end
+        ]
+
+    local_teardown = block_to_fun(teardown_doblock, [:state])
+
+    if length(rest) > 1 do
+      :ok = Logger.warn("#{inspect(mod)}: Only the first teardown is used")
+    end
+
+    opts =
+      case inputs do
+        [] -> opts
+        [_ | _] -> Keyword.put(opts, :inputs, inputs)
+      end
+
+    run_fun_name =
+      ["run", :erlang.unique_integer() |> Integer.to_string(), name]
+      |> Enum.join("_")
+      |> String.to_atom()
+
+    quote generated: true do
+      @unquote(benchee_scenario_attr())(unquote(run_fun_name))
+      def unquote(run_fun_name)(unquote(state_var), opts \\ []) do
+        opts = DeepMerge.deep_merge(opts, unquote(opts))
+
+        case unquote(local_setup).(unquote(state_var)) do
+          {:ok, unquote(state_var)} ->
+            Benchee.run(unquote({:%{}, [], scenarios}), opts)
+            unquote(local_teardown).(unquote(state_var))
+
+          other ->
+            :ok =
+              Logger.error(
+                "Local setup of \"#{unquote(name)}\" in " <>
+                  "#{inspect(__MODULE__)} returned #{inspect(other)}"
+              )
+        end
+      end
+    end
+  end
```

Comment:
The function/macro is almost 100 lines long. It has very distinctive parts, I'd prefer that to be reflected in smaller functions that reflect this (create_setups/create_teardowns, build_benchee_run etc...)

Reply:
Alright!